### PR TITLE
HDDS-4981. Remove SequenceIdGenerator#StateManagerImpl

### PR DESCRIPTION
**Pull request by JacksonYao287**
## What changes were proposed in this pull request?

Remove SequenceIdGenerator#StateManagerImpl

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-4981

## How was this patch tested?

unit test

### CODE REVIEW

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/ha/TestSequenceIDGenerator.java [hunk 1]
```diff
@@ -19,23 +19,26 @@
...
 import org.junit.Test;
 
-import static org.apache.hadoop.hdds.scm.ScmConfigKeys.OZONE_SCM_HA_ENABLE_KEY;
 import static org.apache.hadoop.hdds.scm.ScmConfigKeys.OZONE_SCM_SEQUENCE_ID_BATCH_SIZE;
 
 public class TestSequenceIDGenerator {
   @Test
   public void testSequenceIDGenUponNonRatis() throws Exception {
     OzoneConfiguration conf = SCMTestUtils.getConf();
-    conf.setBoolean(OZONE_SCM_HA_ENABLE_KEY, false);
```
Comment (GlenGeng) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/ha/TestSequenceIDGenerator.java:33]: why remove this configuration ?

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/ha/TestSequenceIDGenerator.java [hunk 2]
```diff
@@ -70,8 +73,6 @@ public void testSequenceIDGenUponNonRatis() throws Exception {
   @Test
   public void testSequenceIDGenUponRatis() throws Exception {
     OzoneConfiguration conf = SCMTestUtils.getConf();
-    // enable ratis based SequenceIDGen
-    conf.setBoolean(OZONE_SCM_HA_ENABLE_KEY, true);
```
Comment (GlenGeng) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/ha/TestSequenceIDGenerator.java:74]: why remove this configuration ?

**Comment by GlenGeng** [comment]
Check `ReconStorageContainerManagerFacade`, in the test, the non ratis id gen can use below `MockSCMHAManager`

```
    this.scmhaManager = MockSCMHAManager.getInstance(
        true, new SCMDBTransactionBufferImpl());
```

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SequenceIdGenerator.java [hunk 3]
```diff
@@ -260,9 +256,13 @@ public Builder setDBTransactionBuffer(DBTransactionBuffer trxBuffer) {
       public StateManager build() {
         Preconditions.checkNotNull(table);
         Preconditions.checkNotNull(buffer);
-        Preconditions.checkNotNull(ratisServer);
 
         final StateManager impl = new StateManagerHAImpl(table, buffer);
+
+        if (null == ratisServer) {
```
[the remark this answers is not in the document]
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SequenceIdGenerator.java:262]: yes, just go ahead here.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/ha/TestSequenceIDGenerator.java (hunk 2 above)
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/ha/TestSequenceIDGenerator.java:74]: before this commit, a sequenceIdGenerator is build depend on this configuration
`
  if (SCMHAUtils.isSCMHAEnabled(conf)) {
      this.stateManager = new StateManagerHAImpl.Builder()
          .setRatisServer(scmhaManager.getRatisServer())
          .setDBTransactionBuffer(scmhaManager.getDBTransactionBuffer())
          .setSequenceIdTable(sequenceIdTable)
          .build();
    } else {
      this.stateManager = new StateManagerImpl(sequenceIdTable);
    }
`
but now ,  it is no use here and a sequenceIdGenerator is build depend on scmHAManager。

we can also keep it here， it does not matter

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/MockSCMHAManager.java [hunk 4]
```diff
@@ -53,6 +53,10 @@ public static SCMHAManager getInstance(boolean isLeader) {
     return new MockSCMHAManager(isLeader);
   }
 
+  public static SCMHAManager getInstance(DBTransactionBuffer buffer) {
```
[the remark this answers is not in the document]
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/MockSCMHAManager.java:56]: MockSCMHAManager.getInstance() will always return a  SCMHAManager instance upon ratis, and now i can not build a sequenctIdGenerator upon no ratis by setting  'OZONE_SCM_HA_ENABLE_KEY'  to false, so the MockSCMHAManager must have a getInstance method which a return a no-ratis SCMHAManager

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManager.java [hunk 5]
```diff
@@ -452,6 +452,11 @@ private void initializeSystemManagers(OzoneConfiguration conf,
       clusterMap = new NetworkTopologyImpl(conf);
     }
 
+    /*
+    TODO: setSCMHAManager is only called in tests now, so
```
[the remark this answers is not in the document]
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManager.java:456]: ok

**Comment by JacksonYao287** [comment]
> Check `ReconStorageContainerManagerFacade`, in the test, the non ratis id gen can use below `MockSCMHAManager`
> 
> ```
>     this.scmhaManager = MockSCMHAManager.getInstance(
>         true, new SCMDBTransactionBufferImpl());
> ```

the code in MockSCMHAManager is below 
```
private MockSCMHAManager(boolean isLeader, DBTransactionBuffer buffer) {
    **this.ratisServer = new MockRatisServer();**// it is not depend on 'isLeader'
    this.isLeader = isLeader;
    this.transactionBuffer = buffer;
  }
```
so , after the above method is called , the 'ratisServer' in MockSCMHAManager will not be null;

the code in SequenceIdGenerator is below
```
      this.stateManager = new StateManagerHAImpl.Builder()
          **.setRatisServer(scmhaManager.getRatisServer())**
          .setDBTransactionBuffer(scmhaManager.getDBTransactionBuffer())
          .setSequenceIdTable(sequenceIdTable)
          .build();
```
```
    final SCMHAInvocationHandler invocationHandler
            **= new SCMHAInvocationHandler(SEQUENCE_ID, impl, ratisServer);**
```
```
  public Object invoke(final Object proxy, final Method method,
                       final Object[] args) throws Throwable {
    try {
      long startTime = Time.monotonicNow();
      final Object result =
          **ratisHandler != null** && method.isAnnotationPresent(Replicate.class) ?
              invokeRatis(method, args) :
              invokeLocal(method, args);
      LOG.debug("Call: {} took {} ms", method, Time.monotonicNow() - startTime);
      return result;
    } catch(InvocationTargetException iEx) {
      throw iEx.getCause();
    }
  }
```
so , we can see that if the 'ratisHandler' is not null and the method is comment with '@Replicate' , invokeRatis will be called。

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/MockSCMHAManager.java [hunk 6]
```diff
@@ -71,6 +76,11 @@ private MockSCMHAManager(boolean isLeader, DBTransactionBuffer buffer) {
     this.transactionBuffer = buffer;
   }
 
+  private MockSCMHAManager(SCMDBTransactionBufferImpl buffer) {
+    this.transactionBuffer = buffer;
```
Comment (GlenGeng) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/MockSCMHAManager.java:80]: The change in `MockSCMHAManager` is not needed. 
The bypass ratis scm is based on the `MockRatisServer`.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/ha/TestSequenceIDGenerator.java [hunk 7]
```diff
@@ -19,23 +19,26 @@
...
 
 public class TestSequenceIDGenerator {
   @Test
   public void testSequenceIDGenUponNonRatis() throws Exception {
     OzoneConfiguration conf = SCMTestUtils.getConf();
-    conf.setBoolean(OZONE_SCM_HA_ENABLE_KEY, false);
     SCMMetadataStore scmMetadataStore = new SCMMetadataStoreImpl(conf);
     scmMetadataStore.start(conf);
 
+    SCMHAManager scmHAManager = MockSCMHAManager.getInstance(
```
Comment (GlenGeng) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/ha/TestSequenceIDGenerator.java:37]:
```
    SCMHAManager scmHAManager = MockSCMHAManager.getInstance(true,
        new SCMDBTransactionBufferImpl());
```

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SequenceIdGenerator.java [hunk 8]
```diff
@@ -91,15 +91,11 @@ public SequenceIdGenerator(ConfigurationSource conf,
...
-          .setSequenceIdTable(sequenceIdTable)
-          .build();
-    } else {
-      this.stateManager = new StateManagerImpl(sequenceIdTable);
-    }
+    Preconditions.checkNotNull(scmhaManager);
+    this.stateManager = new StateManagerHAImpl.Builder()
+        .setRatisServer(scmhaManager.getRatisServer())
+        .setDBTransactionBuffer(scmhaManager.getDBTransactionBuffer())
+        .setSequenceIdTable(sequenceIdTable).build();
```
Comment (GlenGeng) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SequenceIdGenerator.java:98]: Please help rename `StateManagerHAImpl` to `StateManagerImpl`.

**Review by GlenGeng — COMMENTED** [review]
The change looks good. Just some inline comments.

**Comment by GlenGeng** [comment]
> > Check `ReconStorageContainerManagerFacade`, in the test, the non ratis id gen can use below `MockSCMHAManager`
> > ```
> >     this.scmhaManager = MockSCMHAManager.getInstance(
> >         true, new SCMDBTransactionBufferImpl());
> > ```
> 
> @GlenGeng
> the constructor of MockSCMHAManager is below
> 
> ```
> private MockSCMHAManager(boolean isLeader, DBTransactionBuffer buffer) {
>     this.ratisServer = new MockRatisServer();
>     this.isLeader = isLeader;
>     this.transactionBuffer = buffer;
>   }
> ```
> 
> 'this.ratisServer' will always be not null and it is not depend on whether 'isLeader' is true.
> so , after the above method is called , the 'ratisServer' in MockSCMHAManager will not be null.
> 
> the stateManager constructor in SequenceIdGenerator is below
> 
> ```
>       this.stateManager = new StateManagerHAImpl.Builder()
>           .setRatisServer(scmhaManager.getRatisServer())
>           .setDBTransactionBuffer(scmhaManager.getDBTransactionBuffer())
>           .setSequenceIdTable(sequenceIdTable)
>           .build();
> ```
> 
> so here，the ‘ratisServer’ in stateManager will not be null and
> 
> ```
>     final SCMHAInvocationHandler invocationHandler
>             = new SCMHAInvocationHandler(SEQUENCE_ID, impl, ratisServer);
> ```
> 
> so，the ratisHandler below will not be null
> 
> ```
>   public Object invoke(final Object proxy, final Method method,
>                        final Object[] args) throws Throwable {
>     try {
>       long startTime = Time.monotonicNow();
>       final Object result =
>           ratisHandler != null && method.isAnnotationPresent(Replicate.class) ?
>               invokeRatis(method, args) :
>               invokeLocal(method, args);
>       LOG.debug("Call: {} took {} ms", method, Time.monotonicNow() - startTime);
>       return result;
>     } catch(InvocationTargetException iEx) {
>       throw iEx.getCause();
>     }
>   }
> ```
> 
> so , if the 'ratisHandler' is not null and the method is comment with '@replicate' , 'invokeRatis' will be called。
> I modifed MockSCMHAManager so that it can build a instance who's ratisServer is null
> 
> does this make sense?

Hey Jackson, our design decision is, in HA mode, we just `SCMRatisServerImpl`, whose `submitRequest` will call RaftServer; in non-HA mode, or the bypass ratis mode, we use `MockRatisServer`, whose `submitRequest` bypass ratis.

Thus it is fine that the 
>  the 'ratisHandler' is not null and the method is comment with '@replicate' , 'invokeRatis' will be called

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SequenceIdGenerator.java [hunk 9]
```diff
@@ -91,15 +91,11 @@ public SequenceIdGenerator(ConfigurationSource conf,
...
-      this.stateManager = new StateManagerHAImpl.Builder()
-          .setRatisServer(scmhaManager.getRatisServer())
-          .setDBTransactionBuffer(scmhaManager.getDBTransactionBuffer())
-          .setSequenceIdTable(sequenceIdTable)
-          .build();
-    } else {
-      this.stateManager = new StateManagerImpl(sequenceIdTable);
-    }
+    Preconditions.checkNotNull(scmhaManager);
+    this.stateManager = new StateManagerHAImpl.Builder()
```
Comment (GlenGeng) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SequenceIdGenerator.java:95]: enhance java doc `null if non-Ratis based`

**Comment by JacksonYao287** [comment]
> MockRatisServer



> > > Check `ReconStorageContainerManagerFacade`, in the test, the non ratis id gen can use below `MockSCMHAManager`
> > > ```
> > >     this.scmhaManager = MockSCMHAManager.getInstance(
> > >         true, new SCMDBTransactionBufferImpl());
> > > ```
> > 
> > 
> > @GlenGeng
> > the constructor of MockSCMHAManager is below
> > ```
> > private MockSCMHAManager(boolean isLeader, DBTransactionBuffer buffer) {
> >     this.ratisServer = new MockRatisServer();
> >     this.isLeader = isLeader;
> >     this.transactionBuffer = buffer;
> >   }
> > ```
> > 
> > 
> > 'this.ratisServer' will always be not null and it is not depend on whether 'isLeader' is true.
> > so , after the above method is called , the 'ratisServer' in MockSCMHAManager will not be null.
> > the stateManager constructor in SequenceIdGenerator is below
> > ```
> >       this.stateManager = new StateManagerHAImpl.Builder()
> >           .setRatisServer(scmhaManager.getRatisServer())
> >           .setDBTransactionBuffer(scmhaManager.getDBTransactionBuffer())
> >           .setSequenceIdTable(sequenceIdTable)
> >           .build();
> > ```
> > 
> > 
> > so here，the ‘ratisServer’ in stateManager will not be null and
> > ```
> >     final SCMHAInvocationHandler invocationHandler
> >             = new SCMHAInvocationHandler(SEQUENCE_ID, impl, ratisServer);
> > ```
> > 
> > 
> > so，the ratisHandler below will not be null
> > ```
> >   public Object invoke(final Object proxy, final Method method,
> >                        final Object[] args) throws Throwable {
> >     try {
> >       long startTime = Time.monotonicNow();
> >       final Object result =
> >           ratisHandler != null && method.isAnnotationPresent(Replicate.class) ?
> >               invokeRatis(method, args) :
> >               invokeLocal(method, args);
> >       LOG.debug("Call: {} took {} ms", method, Time.monotonicNow() - startTime);
> >       return result;
> >     } catch(InvocationTargetException iEx) {
> >       throw iEx.getCause();
> >     }
> >   }
> > ```
> > 
> > 
> > so , if the 'ratisHandler' is not null and the method is comment with '@replicate' , 'invokeRatis' will be called。
> > I modifed MockSCMHAManager so that it can build a instance who's ratisServer is null
> > does this make sense?
> 
> Hey Jackson, our design decision is, in HA mode, we just `SCMRatisServerImpl`, whose `submitRequest` will call RaftServer; in non-HA mode, or the bypass ratis mode, we use `MockRatisServer`, whose `submitRequest` bypass ratis.
> 
> In both situation, the `@Replicate` annotation, `invokeRatis()` and `submitRequest()` take effect like always, yet `MockRatisServer` just transition the call to be a pure in-mem call.
> 
> Thus it is fine that the
> 
> > the 'ratisHandler' is not null and the method is comment with '@replicate' , 'invokeRatis' will be called

ok，I see ， thanks

**Comment by JacksonYao287** [comment]
@GlenGeng  
please take a review, thanks

**Comment by GlenGeng** [comment]
`org.apache.hadoop.ozone.scm.node.TestDecommissionAndMaintenance ` is a known flaky case. You can retry the CI to see if this case can pass or not.

**Comment by JacksonYao287** [comment]
> `org.apache.hadoop.ozone.scm.node.TestDecommissionAndMaintenance ` is a known flaky case. You can retry the CI to see if this case can pass or not.

ok, thanks , i wiil retry

**Comment by GlenGeng** [comment]
+1. Thanks @JacksonYao287 for the contribution. I will merge it.

**Pull request merged**